src/NuGet.Clients/NuGet.CommandLine/Common/ProtocolDiagnosticsJsonLogger.cs

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Globalization;
+using System.IO;
+using Newtonsoft.Json;
+using NuGet.Protocol.Utility;
+
+namespace NuGet.CommandLine
+{
+    internal sealed class ProtocolDiagnosticsJsonLogger : IDisposable
+    {
+        private object _lock;
+        private StreamWriter _streamWriter;
+        private JsonWriter _jsonWriter;
+
+        internal ProtocolDiagnosticsJsonLogger(string fileName)
+        {
+            _lock = new object();
+            _streamWriter = new StreamWriter(fileName);
+            _jsonWriter = new JsonTextWriter(_streamWriter);
+            _jsonWriter.Formatting = Formatting.Indented;
+            _jsonWriter.WriteStartArray();
+        }
+
+        internal void OnEvent(ProtocolDiagnosticEvent pde)
+        {
+            if (_lock == null)
+            {
+                throw new ObjectDisposedException(nameof(ProtocolDiagnosticsJsonLogger));
+            }
+
+            var timestamp = pde.Timestamp.ToString("o", CultureInfo.InvariantCulture);
+
+            lock (_lock)
+            {
+                _jsonWriter.WriteStartObject();
+
+                _jsonWriter.WritePropertyName("timestamp");
+                _jsonWriter.WriteValue(timestamp);
+
+                _jsonWriter.WritePropertyName("source");
+                _jsonWriter.WriteValue(pde.Source);
+
+                _jsonWriter.WritePropertyName("url");
+                _jsonWriter.WriteValue(pde.Url);
+
+                if (pde.HeaderDuration.HasValue)
+                {
+                    _jsonWriter.WritePropertyName("headerDuration");
+                    _jsonWriter.WriteValue(pde.HeaderDuration.Value.TotalMilliseconds);
+                }
+
+                _jsonWriter.WritePropertyName("duration");
+                _jsonWriter.WriteValue(pde.EventDuration.TotalMilliseconds);
+
+                if (pde.HttpStatusCode.HasValue)
+                {
+                    _jsonWriter.WritePropertyName("httpStatusCode");
```

Comment:
Will we see 401 here in the case of successful auth challenge? e.g. 401, credentials, then 200. Or will we just see the 200?

Also, IIRC this hides redirects under the max redirect count. I think you can sniff for redirect on the [`RequestMessage`](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpresponsemessage.requestmessage?view=netcore-2.0#System_Net_Http_HttpResponseMessage_RequestMessage) property (e.g. different request URI than the original)

Reply:
Just the 200. Again, not worth the effort/tech debt of recording every http request individually.